route-rs-packets/src/arp.rs

```diff
@@ -0,0 +1,253 @@
+    }
+
+    pub fn opcode(&self) -> u16 {
+        let (start, end) = OPCODE_RANGE;
+        u16::from_be_bytes(self.arp_data(start, end).try_into().unwrap())
+    }
+
+    pub fn sender_hardware_addr(&self) -> &[u8] {
+        let (start, end) = self.sender_hardware_addr_range();
+        self.arp_data(start, end)
+    }
+
+    pub fn sender_protocol_addr(&self) -> &[u8] {
+        let (start, end) = self.sender_protocol_addr_range();
+        self.arp_data(start, end)
+    }
+
+    pub fn target_hardware_addr(&self) -> &[u8] {
+        let (start, end) = self.target_hardware_addr_range();
+        self.arp_data(start, end)
+    }
+
+    pub fn target_protocol_addr(&self) -> &[u8] {
+        let (start, end) = self.target_protocol_addr_range();
+        self.arp_data(start, end)
+    }
+
+    pub fn set_hardware_type(&mut self, htype: u16) {
+        let (start, end) = HARDWARE_TYPE_RANGE;
+        self.set_arp_data(&htype.to_be_bytes(), start, end);
+    }
+
+    pub fn set_protocol_type(&mut self, ptype: u16) {
+        let (start, end) = PROTOCOL_TYPE_RANGE;
+        self.set_arp_data(&ptype.to_be_bytes(), start, end);
+    }
+
+    pub fn set_hardware_addr_len(&mut self, len: u8) {
+        let (start, end) = HARDWARE_ADDR_LEN_RANGE;
+        self.set_arp_data(&len.to_be_bytes(), start, end);
+    }
+
+    pub fn set_protocol_addr_len(&mut self, len: u8) {
+        let (start, end) = PROTOCOL_ADDR_LEN_RANGE;
+        self.set_arp_data(&len.to_be_bytes(), start, end);
+    }
+
+    pub fn set_opcode(&mut self, code: u16) {
+        let (start, end) = OPCODE_RANGE;
+        self.set_arp_data(&code.to_be_bytes(), start, end);
+    }
+
+    pub fn set_sender_hardware_addr(&mut self, addr: MacAddr) {
+        // NOTE: should we set len based on frame data, or param?
+        let (start, end) = self.sender_hardware_addr_range();
+        self.set_arp_data(&addr.bytes, start, end);
+    }
+
+    pub fn set_sender_protocol_addr(&mut self, ip_addr: IpAddr) {
+        // NOTE: should we set len based on frame data, or param?
+        let (start, _) = self.sender_protocol_addr_range();
+        self.set_ip_addr(ip_addr, start);
+    }
+
+    pub fn set_target_hardware_addr(&mut self, addr: MacAddr) {
+        // NOTE: should we set len based on frame data, or param?
+        let (start, end) = self.target_hardware_addr_range();
+        self.set_arp_data(&addr.bytes, start, end);
+    }
+
+    pub fn set_target_protocol_addr(&mut self, ip_addr: IpAddr) {
+        // NOTE: should we set len based on frame data, or param?
+        let (start, _) = self.target_protocol_addr_range();
+        self.set_ip_addr(ip_addr, start);
+    }
+
+    // Move ownership of the frame back to the caller
+    pub fn frame(self) -> EthernetFrame {
+        self.frame
+    }
+
+    // Returns the bytes in the ethernet frame between start and end, exclusive
+    fn arp_data(&self, start: usize, end: usize) -> &[u8] {
```

Comment:
You are returning a slice here, which has to have lifetimes. You could also return a COW, then you sidestep the lifetime issues, and may be able to call EthernetFrame's payload() function.